### PR TITLE
Fire Mage

### PR DESCRIPTION
**Pull request by Myrilandel**
# FireMage

## 1. Passive

**PassiveFireMage**
- Passive Spell associated to the FireMage.
- Get in the ConflagrationSpell (Active Spell 3) the current number of IgniteStatus in the game.
- Apply a Critical Bonus Chance = +5% x Number Of Ignite in the game with a maximum of 25% of bonus.

## 2. AutoAttack

**AutoAttackFireMage**
- AutoAttack associated to the FireMage.
- Instantiate a projectile type of AutoAttackFireMage and launch it forward the player.
- Can be notified by the projectile to attribute a Shield to the player.

**AutoAttackFireMageBehaviour**
- Script associated to the Prefab AutoAttackFireMage, instatiated by the auto-attack AutoAttackFireMage.
- If the auto-attack hits a target, it notifies the AutoAttackFireMage script.
- If the auto-attack hits something, the prefab is destroyed

**Shield**
- Script that can be associated to the player when a AutoAttackFireMage prefabs hits a target.
- The shield has for value +5 * Number of Hit with a maximum of +50.
- The shield lasts for 5 seconds or until all of its value is consummed by damages.

## 3. Active Spells
**FireBallSpell**
- Active Spell 1 associated to the FireMage.
- Instantiate a projectile type of FireBall and launch it forward the player.

**FireBall**
- Script associated to the Prefab Fireball, instatiated by the Spell FireBallSpell.
- When the FireBall hits a target, it attaches an IgniteStatus prefab on it and notify the ConflagrationSpell
(Active Spell 2) that an ignite has been added in the game.
- When the FireBall hits something, it is destroyed.

**IgniteStatus**
- Script associated to the Prefab IgniteStatus.
- The IgniteStatus ticks every seconds to apply damage.
- After a MaxDuration, the IgniteStatus is destroyed.

**ConflagrationSpell**
- Active Spell 2 associated to the FireMage.
- Make all IgniteStatus in the game explode.
- If monsters that have no IgniteStatus are in the area of explosion, then, the Ignate status is propagated on them with a amount of chance.

**FireBlessingSpell**
- Active Spell 3 associated to the FireMage.
- Notify the ConflagrationSpell (Active Spell 2) to set a variable call CritSuccess to true.
- When CritSuccess is true, the propagation success rate of exploded IgniteStatus is 100%
- This buff stands during 5 sec OR until the next use of ConflagrationSpell (Active Spell 2)

**SolarBurnSpell**
- Active Spell 4 associated to the FireMage.
- Currently just spawn a FireBall from the Sky to the point you are looking at. 
- This will evolve in a way to applay 4x damages if the damage is a critical hit.
- This will evolve in a way to only launch the spell if the player is targeting the floor.

# KB Tasks & PlantUML 

- [KB Task #63](https://velandelstudio.com/kanboard/?controller=TaskViewController&action=show&task_id=63&project_id=3)
- [KB Task #78](https://velandelstudio.com/kanboard/?controller=TaskViewController&action=show&task_id=78&project_id=3)
- [KB Task #85](https://velandelstudio.com/kanboard/?controller=TaskViewController&action=show&task_id=85&project_id=3)
- [KB Task #106](https://velandelstudio.com/kanboard/?controller=TaskViewController&action=show&task_id=106&project_id=3)
- [UML - Activity Diag](http://www.plantuml.com/plantuml/svg/hLPTJzim57tFhxZIDAamCHLODIc5IULpI6CWolPkDRUjXMjNiSEeqFxxhZ-IsaqAOreVsiO-FZu-FhxEafM66wuad0gczQcI4rmQUCxo2rxYDPkXQnzcxxugSylbGtvBZ_mHenO-XTweCcfa39iyx82PeuJUbVwXiTZ9K1OP2eqMPJbFc10lOKZ2LB1PoQogryZWH239aRDjGB38iHAiX9EgtCe7_fyi-3JxdMNQa2sPNWfkicNZK5WvcTCs8xyY1sFbUGQNQE1g9hb13THzoHOMUA9afU6qv0PEvupiXkDD53p1u7rspPwW2tmmU3zqSgrGAv7D69UHplLUn5BNhSCCo9RULo8TLvC96JYjcYqnPOKx1038HykbMC4PMr0ADDn8k5PI6oo7hfiOndEEeW37EiV2Gcu5Ms7P9hCczwQCllqpm5gO6mw1x5u53Z-Btn8d9m1-Cb4XvF1Xu9gjInizWxxhkKDD6r3JyWMMjWyqJXnrmvoFxMxJxYbHg5-IGFiV0oouJ3_Wmx0b6UCb2h4Vkrqd_RzvxIC2fDLKCU1Vh4wvoF5aiiZoO670nHQ20tnNXazNQV09SILdIPm_mBwbJeNLqxJjFqpCxuouC_mWDZpH45pVF5PVeDzX_VdJKgW2uVmHorLy5E_dwCOvc5zJt4n7qGSjtJkriNVvDrP9CjrDeuCu9wz_TCTaeHP7WpxStBbJ7Z92rFTMO4ki6zSm0SMAia2WifvJTurnem80vapRsUQiI7FGFT0LnpHDRq-K_xoQfl1vMwY2p2QpdPAZxYPsgTqEJisjC7wk2yO6hUS96NqXf8URLQ6viYnfwzhmWsXUwdHbFqq4jIOSmSRZTLtpfGcwwzg1Gm7KEF2IBVhNd0lSk5G3kpisN4CFtGaXYjhNl0k-jzVK4KVgnRoqqDPIBwpBdgL06p1lUBEuLql3fzZrzVNlwmBq_0q3zwoS8PsW2w5Kgv3MvtOzsiQZUV1dpzO2D_elwpn1x5BvX0bBq7NcmbKmAf71pRGfAxkxJvzsTwF2huVPbhHlOMWF3KUXCzcTXRdlyxDzZDyO-WxNPRWlrxlZYjMvm0LAikyvtAJDLlawuak45HbI6JU_4iha9etMx8jucyUdDSfgtbr_icPuKu7sN07Apr2RKgqWrMwTYFAv3gDx7udMbhoWqAfWzFqiVQFPkjGKjdSShTedtC6I2ThBhUNNmiABiD-Otty0)

- [UML - Class Diag](http://www.plantuml.com/plantuml/uml/dLLDRnen43tZNs4f5z1gJpqILKasJOAK1XJIz3piJbYtnaQsbmP5_UyTUvVzX3JA1PWtu-StN-R2EZ2ks4W6A-iCf8wd4gpbYVz4_iOYWeLQyyl2wKld8FqTN0-mmIcBxeLr_BmC-Shj5gMitNRApah_qkELswCDrnNG9L2RvDi9qWEIxxJ8-DAhu-D93_wnpS3X46zu1wvxA5IQ1o5ZeLofPQOoV9smk3piA8jLuR1XUT8XcV4Xv0vK9l7RQufR9xGAH3XXVnahY_OJZ17QlB4phdT8lpDivJ6US7RMnTk2USFp7TPuakHTZa3IQsL8ckXUf0PNMidrz60dXJ6eN426jG-QZzIUP4GdwpiPzSxCLN2KCPZv75PQIpwpRHnM4aFV_95hORSIzi7_e5qxHAz-OLfrCCZhnoS68VId4zcKDyd1hF5GbkD5Zhfm9xs6wKOOl08fAwxebiP-NbxgSgF_UBsq9Gd11dWQlXOa4msXJmQKVT5cmsMPdQcHAHkn-17sALXt9npFXPkoEDGi1tkZ2vKHlD0qW8G_qttQVE7EfzK63EpdHgo58cn1UObKud9l7MxujZ9f7UDZgcUANWQLalfF5gWtXDsLhwVnTZY97XgnqV5fF36Cl1gevfQPVKJ8zeCvxJmHTHyRj5dX3wNzFxfgap_5LJt1vt5yU5ZQa6S-s34rENTnL2uth-0u-e1PMzq6iJBMNzJQbXRQcNpMUz1zmA2axmU5T2vGPVHVz0y0)

_NB : Please Note that some scripts are just in the push because we moved them from a folder to another. These scripts are were already reviewed in other pulls requests. Please do not review th folowing scripts :_
_AutoAttackBase.cs
Character.cs
Spell.cs
FireMage.cs
Warrior.cs_

_NB : Please do not review scripts associated to the Warrior. These scripts are going to be deleted or modified but at the moment they are not used. :_ 
_AutoAttackWarrior.cs
ChargeSpell_

### CODE REVIEW

**Comment by Myrilandel** [comment]
Une bonne grosse release comme on les aime ! 
Release du FireMage. Je vous conseille de bien prendre le temps de lire le résumé de la pull Request pour que vous puissiez reviewer facilement.
Si vous avez des questions n'hésitez pas ! 
Bon courage :D

File: Assets/Scripts/Classes/FireMage/Passive/PassiveFireMage.cs [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+    private float _numberOfIgnites;
+
+    /** Start : protected override void Method
+	 * The Start Method is used here to get the ConflagrationSpell associated to the player and the NumBer Of Ignites in the game.
+	 **/
+    private void Start()
+    {
+        _conflagrationSpell = GetComponent<ConflagrationSpell>();
+        _numberOfIgnites = _conflagrationSpell.Targets.Count;
+    }
```
Comment (Veltouille) [inline Assets/Scripts/Classes/FireMage/Passive/PassiveFireMage.cs:24]: ### Remarque / question : ###

Ce Start(); est il appelé au moment où le mage est chargé ?

Si oui, pourrait on initialiser la valeur de _numberOfIgnites à zéro directement dans les propiétés de la classe ? ou dasn le Start();

Ensuite Update(); le met continuellement à jour ?

File: Assets/Scripts/Classes/FireMage/AutoAttack/AutoAttackFireMage.cs [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+        Invoke("RemoveShield", _maxDurationShield);
+    }
+
+    /** RemoveShield : private void Method
+	 * This method calls the RemoveShield method in the Shield instance.
+	 **/
+    private void RemoveShield()
+    {
+        _shield.RemoveShield();
+    }
```
Comment (Veltouille) [inline Assets/Scripts/Classes/FireMage/AutoAttack/AutoAttackFireMage.cs:96]: ### Commentaire :

Bien, c'est bon pour cette classe pour moi.
Le code est clair et on comprends bien ce que cela fait.
On l'a comprend sans même lire les commentaire xD

Après C'est clair aussi dans le sens ou je vais m'appuyer sur le code pour la refacto des proctiles.

C'est parfait =)

File: Assets/Scripts/Classes/FireMage/ScriptsOnPrefabs/AutoAttackFireMageBehaviour.cs [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+    {
+        EntityLivingBase entityHit = collision.gameObject.GetComponent<EntityLivingBase>();
+        if (entityHit != null && entityHit.gameObject.tag != "Player")
+        {
+            entityHit.DamageFor(5);
+            transform.parent.GetComponent<AutoAttackFireMage>().OnAttackHit();
+        }
+        Destroy(gameObject);
+    }
+}
```
Comment (Veltouille) [inline Assets/Scripts/Classes/FireMage/ScriptsOnPrefabs/AutoAttackFireMageBehaviour.cs:25]: ### Remarque / Question :

Peut être que je me trompe ou que j'ai mal vu, mais je ne vois pas une autodestruction de l'objet si il n'a rien percuté en fonction du temps.

Dit moi tout Jo ^^, sinon il faut le faire car on risque d'avoir des problème de performance !

**Review by Veltouille — COMMENTED** [review]
Hello Jo.

J'ai fait ma review sur les trois premiers scripts de la pull request.

Merci encore pour la qualité de ton travail.

A plus tard =)

File: Assets/Scripts/Classes/FireMage/ScriptsOnPrefabs/AutoAttackFireMageBehaviour.cs (hunk 3 above)
Comment (Myrilandel) [reply Assets/Scripts/Classes/FireMage/ScriptsOnPrefabs/AutoAttackFireMageBehaviour.cs:25]: Oui tu as raison, j'ai déjà répertorié les tâches qui vont apporter cette amélioration (il y a le même problème avec la FireBall) dans le futur.
Je ne les ai pas ajoutés car je pensais les mettre directement dans la classe des Projectiles que tu vas créer.

File: Assets/Scripts/Classes/FireMage/AutoAttack/AutoAttackFireMage.cs (hunk 2 above)
Comment (Myrilandel) [reply Assets/Scripts/Classes/FireMage/AutoAttack/AutoAttackFireMage.cs:96]: Merci beaucoup !

File: Assets/Scripts/Classes/FireMage/Passive/PassiveFireMage.cs (hunk 1 above)
Comment (Myrilandel) [reply Assets/Scripts/Classes/FireMage/Passive/PassiveFireMage.cs:24]: Oui tu as raison merci pour la remarque, je corrigerai ça dans un prochain commit.

File: Assets/Scripts/Classes/FireMage/ScriptsOnPrefabs/AutoAttackFireMageBehaviour.cs (hunk 3 above)
Comment (Veltouille) [reply Assets/Scripts/Classes/FireMage/ScriptsOnPrefabs/AutoAttackFireMageBehaviour.cs:25]: Ok je te remercie de me donner du boulot xD

File: Assets/Scripts/Classes/FireMage/Passive/PassiveFireMage.cs (hunk 1 above)
Comment (Veltouille) [reply Assets/Scripts/Classes/FireMage/Passive/PassiveFireMage.cs:24]: Ok parfait alors !

**Comment by Veltouille** [comment]
A la lecture de tes commentaires, je valide la première partie de la review. je m'y replonge =D

File: Assets/Scripts/Classes/FireMage/ScriptsOnPrefabs/FireBall.cs [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+            }
+            else
+            {
+                ignite = entityHit.gameObject.AddComponent<IgniteStatus>();
+                transform.parent.GetComponent<ConflagrationSpell>().Targets.Add(ignite);
+            }
+        }
+        Destroy(this.gameObject);
+    }
+}
```
Comment (Veltouille) [inline Assets/Scripts/Classes/FireMage/ScriptsOnPrefabs/FireBall.cs:34]: ### Commentaire :

Je n'ai rien  à dire sur cette classe. J'ai tout compris et tout me parait logique.

File: Assets/Scripts/Classes/FireMage/ScriptsOnPrefabs/IgniteStatus.cs [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+
+        _entity = GetComponent<EntityLivingBase>();
+        Invoke("EndStatus", _maxDuration);
+        InvokeRepeating("DamageFor", _tickInterval, _tickInterval);
+    }
+
+    /** DamageFor private void
+	 * This Method is used with the InvokeRepeating Coroutine. That will call the DamageFor() Method in the target.
+	 **/
+    private void DamageFor()
```
Comment (Veltouille) [inline Assets/Scripts/Classes/FireMage/ScriptsOnPrefabs/IgniteStatus.cs:35]: ### Remarque / Question :

Tu as donné à la classe EntityLivingBase et IniteStatus exactement le même nom.
Ne serait-il pas plus clair de renommer la méthode de IgniteStatus et globalement de status (qui infligent des dégâts) en TickDamage(); ??

File: Assets/Scripts/Classes/FireMage/ScriptsOnPrefabs/Shield.cs [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+    {
+        this.ShieldValue += shieldValue;
+    }
+
+    /** ReduceShieldValue public void
+	 * @Param : float value
+	 * This Method is called by other elements in the game to remove points to the shield.
+	 * If the shieldValue reaches 0, the RemoveShield Method is launched.
+	 **/
+    public void ReduceShieldValue(float shieldValue)
```
Comment (Veltouille) [inline Assets/Scripts/Classes/FireMage/ScriptsOnPrefabs/Shield.cs:25]: ### Remarques / Questions :

Que se passe-t-il si la cible à 20 points de Shield et prend 22 dégats ?
Les deux 2 sont-ils appliqués sur les Hp restant de la cible ?

File: Assets/Scripts/Classes/FireMage/ScriptsOnPrefabs/Shield.cs [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+    }
+
+    /** RemoveShield public void
+	 * Instantly Destroy this script.
+	 **/
+    public void RemoveShield()
+    {
+        Destroy(this);
+    }
+}
```
Comment (Veltouille) [inline Assets/Scripts/Classes/FireMage/ScriptsOnPrefabs/Shield.cs:41]: ### Remarque / Question :

De manière plus générale, plutôt qu'attacher un shield et le détruire d'une entité, serait-il selon toi une bonne idée de conserver tout au long de la vie d'une entité (qui vaudrait zéro s'il n'a pas de protection). On pourrait à l'ajout de dégâts, commencer à taper dans le boubou jusqu'à zéro et calculé le restant sur les Hp.

Qu'en penses-tu ?

Sinon tout est toujours clair =)

File: Assets/Scripts/Classes/FireMage/Spells/ConflagrationSpell.cs [hunk 8]
```diff
@@ -0,0 +1,109 @@
+﻿using System.Collections;
+using System.Collections.Generic;
+using UnityEngine;
+
+/** ConflagrationSpell Class, extends Spell
+ * This spell is associated with the FireMageClass
+ * The objectif of this spell is to propagate the IgniteStatus from targets to other targets, making it EXPLODE.
```
Comment (Veltouille) [inline Assets/Scripts/Classes/FireMage/Spells/ConflagrationSpell.cs:7]: ### Commentaires :

**EXPLODE** <-- Lol ...

File: Assets/Scripts/Classes/FireMage/Spells/ConflagrationSpell.cs [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+            Collider[] cols = Physics.OverlapSphere(entity.transform.position, 10f);
+            foreach (Collider col in cols)
+            {
+                if (col.gameObject.GetComponent<EntityLivingBase>()
+                    && !col.isTrigger
+                    && !TargetsExploded.Contains(col)
+                    && !Targets.Contains(col.GetComponent<IgniteStatus>()))
+                {
+                    col.gameObject.GetComponent<EntityLivingBase>().DamageFor(10);
+                    if (Random.Range(0, 100) < 50 || CritSuccess)
```
Comment (Veltouille) [inline Assets/Scripts/Classes/FireMage/Spells/ConflagrationSpell.cs:75]: ### Commentaires / Questions :

Je ne comprends pas bien la ligne du test, peux-tu m'éclairer ?
Existe-t-il par ailleurs un moment ou CritSuccess vaut "true"

Comment (Veltouille) [reply Assets/Scripts/Classes/FireMage/Spells/ConflagrationSpell.cs:75]: J'ai répondu à la seconde question avec FireBlessingSpell...

Mais dans la première condition tu random la range de tes nouvelles ignites ?

File: Assets/Scripts/Classes/FireMage/Spells/SolarBurnSpell.cs [hunk 10]
```diff
@@ -0,0 +1,64 @@
+﻿using System.Collections;
+using System.Collections.Generic;
+using UnityEngine;
+
+/** SolarBurnSpell Class, extends Spell
+ * This spell is associated with the FireMageClass
+ * The objectif of this spell is to Instantiate a prefab (FireBall) and apply a force on it.
+ * This spell also has the particularity to apply quadrupled critical damages.
+ **/
+public class SolarBurnSpell : Spell
```
Comment (Veltouille) [inline Assets/Scripts/Classes/FireMage/Spells/SolarBurnSpell.cs:10]: ### Question :

Tu as codé l'effet du sort ?

**Review by Veltouille — COMMENTED** [review]
Seconde partie et fin de la review.

Je t'ai laissé encore quelques remarques, questions et suggestions. Je te laisse y répondre =)

Tu as fait un sacré boulot Bravo Jo =)

File: Assets/Scripts/Classes/FireMage/ScriptsOnPrefabs/IgniteStatus.cs (hunk 5 above)
Comment (Myrilandel) [reply Assets/Scripts/Classes/FireMage/ScriptsOnPrefabs/IgniteStatus.cs:35]: Oui, ce script va faire l'objet d'une migration vers le concept des Status qui cxontient la méthode TickBehaviour. La tâche est déjà dans le KB :)

File: Assets/Scripts/Classes/FireMage/ScriptsOnPrefabs/Shield.cs (hunk 6 above)
Comment (Myrilandel) [reply Assets/Scripts/Classes/FireMage/ScriptsOnPrefabs/Shield.cs:25]: Pour l'instant le systeme de shield n'a pas été intégré avec le système de HP. J'attendais la release de Fab avec les barres de ressources pour adapter cela. Mais concrètement oui ce serait l'idée on absorbe X points de degats sur le shield et le restant est appliqué sur les HP

File: Assets/Scripts/Classes/FireMage/ScriptsOnPrefabs/IgniteStatus.cs (hunk 5 above)
Comment (Veltouille) [reply Assets/Scripts/Classes/FireMage/ScriptsOnPrefabs/IgniteStatus.cs:35]: OK, super !

File: Assets/Scripts/Classes/FireMage/ScriptsOnPrefabs/Shield.cs (hunk 6 above)
Comment (Veltouille) [reply Assets/Scripts/Classes/FireMage/ScriptsOnPrefabs/Shield.cs:25]: Du coup Cf la seconde remarque sur la class !

File: Assets/Scripts/Classes/FireMage/Spells/ConflagrationSpell.cs (hunk 8 above)
Comment (Myrilandel) [reply Assets/Scripts/Classes/FireMage/Spells/ConflagrationSpell.cs:7]: C'est pour marquer la violence

File: Assets/Scripts/Classes/FireMage/ScriptsOnPrefabs/Shield.cs (hunk 7 above)
Comment (Myrilandel) [reply Assets/Scripts/Classes/FireMage/ScriptsOnPrefabs/Shield.cs:41]: On peut effectivement faire ça c'est une bonne idée. On s'est pas trop penché sur l'architecture des boucliers, je pense qu'il feront l'objet d'une refonte eux aussi avec une architecture plus adéquat à leur utilisation et on pourra effectivement faire comme ça.

File: Assets/Scripts/Classes/FireMage/Spells/ConflagrationSpell.cs (hunk 9 above)
Comment (Myrilandel) [reply Assets/Scripts/Classes/FireMage/Spells/ConflagrationSpell.cs:75]: Non on random les chances de success. En gros, l'ignite Explose et inflige des dégâts à tous les monstres à portée. Si ces monstres n'avaient pas d'ignites alors ils ont 50% de chance d'en obtenir une. La propagation est donc pas assurée sauf si on active la Fire Blessing

File: Assets/Scripts/Classes/FireMage/Spells/SolarBurnSpell.cs (hunk 10 above)
Comment (Myrilandel) [reply Assets/Scripts/Classes/FireMage/Spells/SolarBurnSpell.cs:10]: Non car on ne gère pas encore via un script les dégâts critique

File: Assets/Scripts/Classes/FireMage/Spells/ConflagrationSpell.cs (hunk 9 above)
Comment (Veltouille) [reply Assets/Scripts/Classes/FireMage/Spells/ConflagrationSpell.cs:75]: Ok ça marche merci =)

**Review by Veltouille — APPROVED** [review]
Et bien je n'ai pas d'autre remarque, tu as répondu à toutes mes questions c'est parfait !

Encore un super super boulot !

On va pouvoir créer de nouveaux axes de travail =)

**Comment by Myrilandel** [comment]
Je pense que je vais merger. Pour l'amélioration du passif, j'ai fais une tâche sur le KB on fera migrer tout ça dans une prochaine release !
Merci pour tes remarques :)

**Comment by Veltouille** [comment]
De rien bro ! Tu as beaucoup bossé et il y avait pas mal a lire ! Mais c'était super intéressant !

J'ai hâte de m'essayer à faire mon Character =)

Mergeons :+1:

**Pull request merged**